File: corelibrary/src/main/java/pers/cxd/corelibrary/EventHandler.java
package pers.cxd.corelibrary;

import android.os.Handler;
import android.os.Looper;
import android.os.Message;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * EventHandler is designed to send event between {@link pers.cxd.corelibrary.base.UiComponent}
 *
 * @author pslilysm
 * @since 1.0.0
 */
public class EventHandler extends Handler {

    private static final Class<?>[] sConstructorClasses = new Class[]{Looper.class};
    private final ConcurrentMap<Integer, List<EventCallback>> mMultiCallbacks = new ConcurrentHashMap<>();

    public EventHandler(@NonNull Looper looper) {
        this(looper, null);
    }

    public EventHandler(@NonNull Looper looper, @Nullable Handler.Callback callback) {
        super(looper, callback);
    }

    /**
     * @return a default EventHandler which bind MainLooper;
     */
    public static EventHandler getDefault() {
        return SingletonFactory.findOrCreate(EventHandler.class, sConstructorClasses, Looper.getMainLooper());
    }

    public void registerEvent(int eventCode, EventCallback callback) {
        mMultiCallbacks.computeIfAbsent(eventCode, i -> new CopyOnWriteArrayList<>()).add(callback);
    }

    public void unregisterEvent(int eventCode, EventCallback callback) {
        mMultiCallbacks.getOrDefault(eventCode, Collections.emptyList()).removeIf(eventCallback -> eventCallback == callback);
    }

    public void unregisterAllEvent(EventCallback callback) {
        mMultiCallbacks.values().forEach(eventCallbacks -> eventCallbacks.remove(callback));
    }

    /**
     * if current thread == the thread we bind, handle the event directly
     *
     * @param event to send
     */
    public void sendEventOpted(Message event) {
        if (Looper.myLooper() == getLooper()) {
            handleMessage(event);
            event.recycle();
        } else {
            sendMessage(event);
        }
    }

    @Override
    public void handleMessage(@NonNull Message msg) {
        List<EventCallback> callbacks = mMultiCallbacks.getOrDefault(msg.what, Collections.emptyList());
        callbacks.forEach(eventCallback -> eventCallback.handleEvent(msg));
    }

    public interface EventCallback {
        void handleEvent(@NonNull Message msg);
    }

}

File: corelibrary/src/main/java/pers/cxd/corelibrary/EventHandler.kt
package pers.cxd.corelibrary

import android.os.Handler
import android.os.Looper
import android.os.Message
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentMap
import java.util.concurrent.CopyOnWriteArrayList
import java.util.function.Consumer

/**
 * EventHandler is designed to send event between [pers.cxd.corelibrary.base.UIComponent]
 *
 * @author pslilysm
 * @since 1.0.0
 */
class EventHandler @JvmOverloads constructor(looper: Looper, callback: Callback? = null) :
    Handler(looper, callback) {
    private val mMultiCallbacks: ConcurrentMap<Int, MutableList<EventCallback>> =
        ConcurrentHashMap()

    private val mEmptyMutableList: MutableList<EventCallback> = mutableListOf()

    fun registerEvent(eventCode: Int, callback: EventCallback) {
        mMultiCallbacks.computeIfAbsent(eventCode) { CopyOnWriteArrayList() }
            .add(callback)
    }

    fun unregisterEvent(eventCode: Int, callback: EventCallback) {
        mMultiCallbacks.getOrDefault(eventCode, mEmptyMutableList)
            .removeIf { eventCallback: EventCallback -> eventCallback === callback }
    }

    fun unregisterAllEvent(callback: EventCallback) {
        mMultiCallbacks.values.forEach(Consumer { eventCallbacks: MutableList<EventCallback> ->
            eventCallbacks.remove(
                callback
            )
        })
    }

    /**
     * if current thread == the thread we bind, handle the event directly
     *
     * @param event to send
     */
    fun sendEventOpted(event: Message) {
        if (Looper.myLooper() == looper) {
            handleMessage(event)
            event.recycle()
        } else {
            sendMessage(event)
        }
    }

    override fun handleMessage(msg: Message) {
        mMultiCallbacks.getOrDefault(msg.what, mEmptyMutableList)
            .forEach(Consumer { eventCallback: EventCallback -> eventCallback.handleEvent(msg) })
    }

    interface EventCallback {
        fun handleEvent(msg: Message)
    }

    companion object {
        private val sDefault = object : Singleton<EventHandler>() {
            override fun create(): EventHandler {
                return EventHandler(Looper.getMainLooper())
            }
        }

        /**
         * @return a default EventHandler which bind MainLooper;
         */
        @kotlin.jvm.JvmStatic
        val default: EventHandler
            get() = sDefault.getInstance()
    }
}